### PR TITLE
E2e/script+profiler

### PR DESCRIPTION
**Pull request by ronyw7**
### Changes
`ray_data_e2e_training.py`
- Refactored preprocessing pipeline, it now uses custom-versioned Ray which comes with the `transform=` parameter in `read_images` API. Also uses a custom `collate_fn` to for normalizing and converting numpy image arrays to torch.tensors.
- Added a custom profiler that correctly shows the timeline of GPU events

`create_sample_dataset.py`
- Minor fix, `../ILSVRC/Data/CLS-LOC/val/` contains image folders, not images.

### CODE REVIEW

File: ray_data_eval/image_training/create_sample_dataset.py [hunk 1]
```diff
@@ -75,9 +75,10 @@ def copy_random_folders(source_dir, target_dir, percentage, copy_dir=True):
     args = parser.parse_args()
 
     PERCENTAGE = args.percentage
-    SOURCE_DIR = "/home/ubuntu/image-data/ILSVRC/Data/CLS-LOC/train/"
+    SYS = 'azureuser'
```
Comment (franklsf95) [inline ray_data_eval/image_training/create_sample_dataset.py:78]: maybe use `USERNAME`

File: ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py [hunk 2]
```diff
@@ -22,7 +22,7 @@
 import torchvision.transforms as transforms
 from torch.optim.lr_scheduler import StepLR
 from ray.data.datasource.partitioning import Partitioning
-import numpy as np
+# import numpy as np
```
Comment (franklsf95) [inline ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py:25]: remove if not needed

File: ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py [hunk 3]
```diff
@@ -86,6 +86,11 @@
 
 def main():
     args = parser.parse_args()
+    local = True if (args.data[0] == '/' or args.data[0] == '~') else False
```
Comment (franklsf95) [inline ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py:89]: maybe `args.data.startswith("s3://")`?

Also, instead of saying `x = True if (a == b) else False`, just say `x = a == b`

File: ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py [hunk 4]
```diff
@@ -325,40 +327,103 @@ def train(train_dataset, model, criterion, optimizer, epoch, device, args):
 
     i = 0
     end = time.time()
-    for batch in train_dataset.iter_torch_batches(batch_size=args.batch_size):
+
+    def _collate_fn_arr_pil_images_to_tensor(batch:dict) -> dict:
```
Comment (franklsf95) [inline ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py:331]: space after `:`. did you run formatter?

**Review by franklsf95 — APPROVED** [review]
Very nice! Some inline comments then let's merge it.

File: ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py (hunk 3 above)
Comment (ronyw7) [reply ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py:89]: thanks! `args.data.startswith("s3://")` is way better

File: ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py (hunk 4 above)
Comment (ronyw7) [reply ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py]: Oops, turned out I forgot to install Ruff on my azure machine. Looked like the previous commits passed the push checks but I just reformatted the documents using Ruff.

**Comment by ronyw7** [comment]
I've resolved the issues and added code for auto-combining CPU and GPU timeline logs, which simplifies the workflow (i.e. no manual copying of json contents). Let me know if this is good to merge!

File: ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py [hunk 5]
```diff
@@ -125,6 +189,27 @@ def main():
         # Simply call main_worker function
         main_worker(args.gpu, ngpus_per_node, args)
 
+    ray.timeline(timeline_filename)
+
+    # Append gpu event log to the main log
```
Comment (franklsf95) [inline ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py:194]: Very nice! Make a separate function for this.

File: ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py [hunk 6]
```diff
@@ -304,71 +386,141 @@ def val_transform(row):
...
+        tensor_batch = tensor_batch.permute(0, 3, 1, 2)
+        tensor_batch = normalize(tensor_batch)
+
+        labels = torch.tensor(batch["label"], dtype=torch.long)
+        batch["image"] = tensor_batch
+        batch["label"] = labels
+
+        return batch
+
+    class ChromeTracer:
```
Comment (franklsf95) [inline ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py:429]: Can we define this at top level instead of in this train function?

File: ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py [hunk 7]
```diff
@@ -304,71 +386,141 @@ def val_transform(row):
...
+            self._add_event(self.name, "E", self.end_time)
+
+        def profile(self, name, extra_data=None):
+            self.name = name
+            self.extra_data = extra_data
+            return self
+
+        def save(self):
+            with open(self.log_file, "w") as f:
+                import json
```
Comment (franklsf95) [inline ray_data_eval/image_training/e2e_training/ray_data_e2e_training.py:468]: Can we just `import json` at top level?

**Review by franklsf95 — APPROVED** [review]
Some inline comments, then good to merge!
